lib/LaTeXML/Core/Document.pm

```diff
@@ -1439,7 +1439,63 @@ sub pruneXMDuals {
       $self->collapseXMDual($dual, $presentation); }
     elsif (!$self->findnode('descendant-or-self::*[@_pvis or @_cvis]', $presentation)) {    # pres.
       $self->collapseXMDual($dual, $content); }
-  }
+    else {    # compact aligned structures, where possible
+      $self->compactXMDual($dual, $content, $presentation); } }
+  return; }
+
+sub compactXMDual {
+  my ($self, $dual, $content, $presentation) = @_;
+  # For now only handle compacting mirror applications
+  return if ($self->getNodeQName($content) ne 'ltx:XMApp') ||
+    ($self->getNodeQName($presentation) ne 'ltx:XMApp');
+  my @content_args = element_nodes($content);
+  my @pres_args    = element_nodes($presentation);
+  return if scalar(@content_args) != scalar(@pres_args);
+
+  my @new_args = ();
+  my $single_duality;
+  # walk the corresponding children, and double-check they are referenced in the same order
+  while ((my $c_arg = shift(@content_args)) and (my $p_arg = shift(@pres_args))) {
+    my $c_idref = $c_arg->getAttribute('idref');
+    if ($c_idref && ($c_idref eq ($p_arg->getAttribute('xml:id') || ''))) {
+      push @new_args, $p_arg;
+      next; }    # content-refs-pres, OK
+    my $p_idref = $p_arg->getAttribute('idref');
+    if ($p_idref && ($p_idref eq ($c_arg->getAttribute('xml:id') || ''))) {
+      push @new_args, $c_arg;
+      next; }    # pres-refs-content, OK
+
+    # Next up: differences. If:
+    # 1) we saw such a difference beforehand, or
+    # 2) the tree is too complex - give up on compacting and return.
+    # we only handle two XMToks differing for now.
+    if ($single_duality || ($self->getNodeQName($c_arg) ne 'ltx:XMTok') || $self->getNodeQName($p_arg) ne 'ltx:XMTok') {
```

Comment:
why does parg need to be a token? (so long as it's consistently id/ref'd)

Reply:
It was just a starting point. And I am not sure if passing the meaning attribute to other types of nodes wouldn't confuse the post-processor, so I'd like some examples... E.g. XMArray, XMRow, XMCell, or even another XMDual nested underneath could cause some hiccups. But maybe they're all workable already, haven't tried.

Reply:
You don't need examples; I need counter-examples!

I think you're looking at it a bit backwards; The attributes *inform* post-processor, if that makes it confused, that's the post-processor's problem.  Or alternatively, assume the attributes are there for a reason, possibly bad, but they don't need to justify themselves.

Reply:
Alright, I'll transfer to arbitrary nodes then, scary as that seems... Just landed a commit that removes this guard, though I still haven't found a good example of what it does. Should probably rebase the a11y branch on this one and see how things change in the showcase...